feat(cmd): Add inspect block subcommand to inspect blocks

kostko opened the pull request:
Example output:
```
$ oasis inspect block 429810 0
Network:        mainnet
ParaTime:       sapphire (Sapphire Mainnet)
Round:          429810
Version:        0
Namespace:      000000000000000000000000000000000000000000000000f80306c9858e7279
Timestamp:      2023-05-10T13:58:31+02:00
Type:           1
Previous:       27b533249293dbd89c201afb0e7016f374227dece392a244d2524fc0c65dd9ba
I/O root:       3aab20006fd571491f9b2ddff140a4837b6777a5cc1823cc8c00e68359db0b29
State root:     3e9d977ddb72538aa5febfaf7e8e8bdb8d02fcb555511e699cd2ffa26a802efb
Messages (out): 65f1a41bea764611d33e0374ddc927e1e1696fb282ba8b81779730b0a32dccc6
Messages (in):  c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
Transacations:  1

=== Transaction 0 ===
Kind: oasis
Hash: e2049ba3f28e5937606048d00267a3f2adde1d3ca5f646b360ecad4767bf8801
Signer(s):
  1. Ax5Y8e9iqJv/fndH+pd6Xi+LiONEPY6DtMg0fXcDauld
     (signature: MEUCIQCzFdZUyC3dyzRrdb+CMZz6QmxM8BV4vrsf1C/G3s2KiQIgRqHdTzN1P+LWDovqhkB3cdYhFaMKsei/iPzTqz+F+MA=)
Content:
  Format: plain
  Method: consensus.Withdraw
  Body:
    To: oasis1qry5d5zxs58w9deu5ara5zaamhkzqtg99s00ls90
    Amount: 0.1 ROSE
  Authorized signer(s):
    1. Ax5Y8e9iqJv/fndH+pd6Xi+LiONEPY6DtMg0fXcDauld (secp256k1eth)
       Nonce: 71079
  Fee:
    Amount: 0.0011321 ROSE
    Gas limit: 11321
    (gas price: 0.0000001 ROSE per gas unit)

=== Result of transaction 0 ===
Status: ok
Data:
  null

=== Events emitted by transaction 0 ===
Events: 2

  --- Event 0 ---
  Module: accounts
  Code:   1
  Data:
    - transfer:
        from: oasis1qquzek9y9guk6plys32s7zvf4g2f8rvn9slu8uye
        to: oasis1qr677rv0dcnh7ys4yanlynysvnjtk9gnsyhvm6ln
        amount:
          amount: "100000000000000000"
          denomination: ""
      burn: null
      mint: null

  --- Event 1 ---
  Module: core
  Code:   1
  Data:
    - amount: 11315
```

Or for Ethereum transactions:
```
$ oasis inspect block 429809 0
Network:        mainnet
ParaTime:       sapphire (Sapphire Mainnet)
Round:          429809
Version:        0
Namespace:      000000000000000000000000000000000000000000000000f80306c9858e7279
Timestamp:      2023-05-10T13:57:27+02:00
Type:           1
Previous:       72b9184af6f7c3e9760df2b332e3faf28bccaf65df3127c379187719bdb42805
I/O root:       084e4a1e14ef1886aa92c5affb57014487c6cd0538573f6030ee10877e4c5529
State root:     e3b4e31328d7148b47281ebd46ab005c503e714f281c73b0594e229522ca730d
Messages (out): c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
Messages (in):  c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
Transacations:  1

=== Transaction 0 ===
Kind:      evm.ethereum.v0
Hash:      cfe71943945165c9505bc55d6dd9ce41fecacf17b4bccb13d1d11de6d02147c1
Eth hash:  0x1eb0af040d468a861807796a7a7577770619142dd01857d5b1625efb0decbf2d
Chain ID:  23294
Nonce:     67196
Type:      0
To:        0xCCD706B1c30d1c6F46F74885C3B13e3Ea28ce8AF
Value:     100000001683719844
Gas limit: 30000
Gas price: 100000000000
Data:
  (none)

=== Result of transaction 0 ===
Status: ok
Data:
  []

=== Events emitted by transaction 0 ===
Events: 1

  --- Event 0 ---
  Module: core
  Code:   1
  Data:
    - amount: 22143
```

matevz commented on cmd/inspect/block.go:
```diff
@@ -0,0 +1,219 @@
+
+var blockCmd = &cobra.Command{
+	Use:   "block <number> [ <tx-index> | <tx-hash> ]",
+	Short: "Show information about a block and its transactions",
```
```suggestion
	Short: "Show information about a block and its transactions",
	Long: "Show information about a given block number and (optionally) its transactions. Use \"latest\" to use the last minted block.",
```

matevz commented on cmd/inspect/block.go:
```diff
@@ -0,0 +1,219 @@
+						}
+
+						fmt.Printf("Kind:      evm.ethereum.v0\n")
+						fmt.Printf("Hash:      %s\n", ethTx.Hash())
```
Similar to the "Address" vs "Ethereum address", should we discriminate between oasis and eth hash?

```suggestion
						fmt.Printf("Ethereum hash:      %s\n", ethTx.Hash())
						fmt.Printf("Hash:      %s\n", tx.Tx.Hash())
```

matevz reviewed (changes requested):
Do we show emitted events of the transaction anywhere?

kostko commented:
Fixed the above and also added support for events (it would be even nicer if we supported pretty printing for events, but this will need to be changed in the SDK and can be added later).

matevz commented on cmd/inspect/block.go:
```diff
@@ -0,0 +1,267 @@
+			txs, err := rt.GetTransactionsWithResults(ctx, blk.Header.Round)
+			cobra.CheckErr(err)
+
+			fmt.Printf("Transacations:  %d\n", len(txs))
```
```suggestion
			fmt.Printf("Transactions:  %d\n", len(txs))
```

matevz commented on cmd/inspect/block.go:
```diff
@@ -0,0 +1,267 @@
+	fmt.Println(indent + output)
+}
+
+func prettyPrintEvent(indent string, evIndex int, ev *types.Event, decoders []client.EventDecoder) {
```
Fields like `address`, `topics` and `data` in events are currently printed as a list of integers. Could you fix it to be an ordinary hex-encoded value. That's the standard fallback for EVM tools I think. Is this the pretty printing you were referring to in https://github.com/oasisprotocol/cli/pull/72#issuecomment-1543584935 ?

```
$ oasis inspect block 5653027 3 --network mainnet   --paratime emerald
...
  --- Event 1 ---
  Module: evm
  Code:   1
  Data:
    - address:
      - 240
      - 43
      - 62
      - 67
      - 115
      - 4
      - 137
      - 33
      - 5
      - 153
      - 37
      - 18
      - 83
      - 159
      - 118
      - 148
      - 35
      - 165
      - 21
      - 203
      topics:
      - - 221
        - 242
        - 82
        - 173
        - 27
        - 226
        - 200
        - 155
        - 105
        - 194
        - 176
        - 104
        - 252
...
```

kostko replied:
This cannot be done without changing the SDK and should be handled as we add pretty printing there.
